rustls/src/conn/unbuffered.rs

```diff
@@ -18,6 +20,10 @@ impl<Data> UnbufferedConnectionCommon<Data> {
         let mut buffer = MessageDeframerSliceBuffer::new(incoming_tls, &mut discard);
 
         let state = 'outer: loop {
+            if let Some(chunk) = self.pop_early_data() {
```

Comment:
Is the dependency on RTTI introduced in `pop_early_data` better handled by having a common (to clients and servers) `process_tls_records()` that is called by a client- and server-specific version? This clause could then exist only in the server version.

Reply:
that worked. however, now the tests won't compile because they were relying on the generic `UnbufferedConnectionCommon<Data>` having the `process_tls_records` method. I basically did this
``` rust
impl UnbufferedConnectionCommon<ClientConnectionData> {
    pub fn process_tls_records<>(/* .. */) -> /* .. */ {
        // call common code
    }
}

impl UnbufferedConnectionCommon<ServerConnectionData> {
    pub fn process_tls_records<>(/* .. */) -> /* .. */ {
        // perform server-only logic
        // call common code
    }
}
```
and now I cannot add this because of the name collision
``` rust
impl<Data> UnbufferedConnectionCommon<Data> {
    pub fn process_tls_records<>(/* .. */) -> /* .. */ {
        // call common code
    }
}
```
however, I could add that last method to the generic type under a different name ...

another option would be to add an `UnbufferedConnection` enum that runtime-represents both the client and server sides and have the `process_tls_records` method there. the tests could then use that `enum` but that would result in several changes in the test code

Reply:
> now the tests won't compile because they were relying on the generic UnbufferedConnectionCommon<Data> having the process_tls_records method

fixing up the tests to not rely on the generic struct having that method turned out to be not too onerous. I didn't have to add the `enum`.